Implement card-based visual layout and other visual changes

hbillings opened the pull request:
Relies on #1066.

Implementing @ericronne's designs one piece at a time. This constrains the width of the forms to 6 of the available 12 columns (with an extra column of padding on either side). 

Screenshots will update here as applicable.
Logged-out homepage. I'd really like to add the log in link to this new, less prominent nav:
<img width="972" alt="screen shot 2016-12-15 at 9 25 09 am" src="https://cloud.githubusercontent.com/assets/509309/21232326/684b0968-c2a8-11e6-8c87-99a3827ff23a.png">

Logged-in homepage header:
<img width="957" alt="screen shot 2016-12-15 at 9 07 56 am" src="https://cloud.githubusercontent.com/assets/509309/21231615/26192cac-c2a6-11e6-9847-41b757a6a8ab.png">

Narrow card view, with header:
![screencapture-0-0-0-0-8000-data-capture-step-1-1481819148057](https://cloud.githubusercontent.com/assets/509309/21232364/8ab42fb6-c2a8-11e6-860e-3ff3cbec67db.png)

Wide card view, no header:
![screencapture-0-0-0-0-8000-data-capture-step-3-errors-1481819805913](https://cloud.githubusercontent.com/assets/509309/21232769/1325eb4a-c2aa-11e6-8797-4d314ef75d9f.png)

Narrow card view, no header:
<img width="973" alt="screen shot 2016-12-15 at 9 23 21 am" src="https://cloud.githubusercontent.com/assets/509309/21232227/2efd7876-c2a8-11e6-9986-471c33e6e022.png">


To do:
- [x] Get Step 2 to reflect the schedule name
- [x] Update site admin page styles

ericronne commented:
sweet!

hbillings commented:
Site admin page:
![screencapture-0-0-0-0-8000-admin-1481148482710](https://cloud.githubusercontent.com/assets/509309/20988692/06ca6482-bc8f-11e6-9922-f94fe98ec2ae.png)

hbillings commented on data_capture/views/price_list_upload.py:
```diff
@@ -175,10 +177,11 @@ def step_2(request, step):
     return step.render(request, {
         'form': form,
-        'step_1_data': get_step_form_from_session(1, request)
+        'preferred_schedule': step_1_data.cleaned_data['schedule_class'],
```
This isn't the right thing -- this gives me `data_capture.schedules.s70.Schedule70PriceList`. Not sure what I need to do to get the name of the schedule to show up.

jseppi replied:
Maybe `step_1_data.cleaned_data['schedule_class'].title` ? (like at https://github.com/18F/calc/blob/8c5287ee4962f17cc76242cfeba301ad6952655d/data_capture/views/price_list_upload.py#L405)
I had trouble with this previously also :(

hbillings replied:
@jseppi Ah!!! That was it. Thank you!

hbillings commented:
@ericronne Because we moved the vendor name from step 1 to step 2, I moved around some of the info on step 2. It looks like this now (and could probably use some help):
<img width="602" alt="screen shot 2016-12-08 at 9 48 46 am" src="https://cloud.githubusercontent.com/assets/509309/21018878/d012d468-bd2b-11e6-9a55-789c9687fb9c.png">

jseppi commented on data_capture/views/price_list_upload.py:
```diff
@@ -154,10 +156,10 @@ def step_2(request, step):
     return step.render(request, {
         'form': form,
-        'step_1_data': get_step_form_from_session(1, request)
+        'preferred_schedule': step_1_data.cleaned_data['schedule_class'].title,
```
Rename this to `'preferred_schedule_title'` since that seems to be how it is differentiated from the class name in other other view methods here.

jseppi commented on data_capture/templates/data_capture/price_list/step_2.html:
```diff
@@ -35,9 +34,11 @@
+
+{% endblock %}
+{% block card_footer %}
+  <button type="submit" class="button-link" name="cancel" data-a11y-dialog-show="cancel-dialog" formnovalidate>Cancel this upload</button>
```
Hrm, do all these `cancel` buttons still work? I think they might have to be inside the `<form>`.

jseppi replied:
Just checked. They work when JS works because the modal dialog has another `<form>` in it. I _think_ you should be able to just wrap each of these cancel buttons in yet another `<form>`:

```html
<form method="post">
  {% csrf_token %}
  <button type="submit" class="button-link" name="cancel" data-a11y-dialog-show="cancel-dialog" formnovalidate>Cancel this upload</button>
</form>

jseppi commented on data_capture/templates/data_capture/price_list/step_3.html:
```diff
@@ -41,5 +33,10 @@ <h2 id="upload" tabindex="-1">Want to try that again?</h2>
 {% endblock %}
+{% block card_footer %}
+  <button type="submit" class="button-link" name="cancel" data-a11y-dialog-show="cancel-dialog" formnovalidate>Cancel this upload</button>
+</div>
```
dangling `</div>` ?

jseppi replied:
Or maybe it isn't because of the upload form partial?

jseppi commented on data_capture/templates/data_capture/price_list/step_3_errors.html:
```diff
@@ -1,24 +1,22 @@
           </p>
         {% else %}
-          <h3>{{ total }} row{{ total|pluralize:" has,s have" }} errors</h3>
+          <h3>We found errors in {{ total }} row{{ total|pluralize:",s" }} of your data</h3>
```
I think you can leave off the `:",s"` in the puralize filter since that is what it defaults to adding.

jseppi commented on data_capture/templates/data_capture/price_list/step_2.html:
```diff
@@ -1,25 +1,24 @@
   {% fieldset form.contractor_site %}
 
   <div class="date-range {% if form.contract_start.errors or form.contract_end.errors%}fieldset-error{% endif %}">
     {% fieldset form.contract_start %}
```
Spacing between this fieldset and the previous one seems off:

<img width="325" alt="screen shot 2016-12-08 at 1 45 44 pm" src="https://cloud.githubusercontent.com/assets/697848/21025083/b5e678c6-bd4c-11e6-9d43-cf42eb0d6752.png">

hbillings commented:
Okay, @jseppi, I think I hit everything! Also threw in a fix for #1088, and made one small tweak to the design. I realized I missed the fact that the cards were supposed to "stack" on top of the header background, like so: 
<img width="966" alt="screen shot 2016-12-08 at 5 31 28 pm" src="https://cloud.githubusercontent.com/assets/509309/21033243/65bf0988-bd6c-11e6-8949-5fa989af7aab.png">

jseppi commented:
Hrm, overlap looks like it was missed in a few places:

Front page:
<img width="709" alt="screen shot 2016-12-09 at 8 14 28 am" src="https://cloud.githubusercontent.com/assets/697848/21051819/90be4b46-bde7-11e6-83cb-44a1ac02e74c.png">

Account page - no price lists added yet (`/account/`):
<img width="975" alt="screen shot 2016-12-09 at 8 15 13 am" src="https://cloud.githubusercontent.com/assets/697848/21051962/407fc424-bde8-11e6-8a87-d4b43aa48983.png">

Account page - with price lists (`/account/`):
<img width="972" alt="screen shot 2016-12-09 at 8 18 58 am" src="https://cloud.githubusercontent.com/assets/697848/21052000/68f46072-bde8-11e6-8aa9-19a1767985f8.png">

Admin page (`/admin/`)
<img width="958" alt="screen shot 2016-12-09 at 8 19 16 am" src="https://cloud.githubusercontent.com/assets/697848/21052005/6eac156e-bde8-11e6-870b-4f3f0db87561.png">

Price List Details page (`/data-capture/price-lists/1`)
<img width="966" alt="screen shot 2016-12-09 at 8 19 03 am" src="https://cloud.githubusercontent.com/assets/697848/21052029/8d77397e-bde8-11e6-9516-ca1bd1fffcb7.png">

Price List List page (`/data-capture/price-lists/`)
<img width="806" alt="screen shot 2016-12-09 at 8 22 13 am" src="https://cloud.githubusercontent.com/assets/697848/21052046/b0639dec-bde8-11e6-98f3-f6493f363064.png">

jseppi commented:
Also on the Step 4 in-page edit form, the start/end date fields are still "inline" unlike in Step 2 (maybe this is intentional):

<img width="779" alt="screen shot 2016-12-09 at 8 16 32 am" src="https://cloud.githubusercontent.com/assets/697848/21052078/cfa34b12-bde8-11e6-9505-044ab868b109.png">

jseppi commented:
eeep, and "Step 3 Data" (`step_3_data.html`), the one you get to if finish the upload in Step 3, but then go back to Step 2 and forward to Step 3 again:

<img width="1027" alt="screen shot 2016-12-09 at 8 25 29 am" src="https://cloud.githubusercontent.com/assets/697848/21052180/3261600e-bde9-11e6-9346-e12b139d8d27.png">

jseppi commented on data_capture/templates/data_capture/price_list/step_2.html:
```diff
@@ -35,9 +34,13 @@
+
+{% endblock %}
+{% block card_footer %}
+  <form>
```
All of these "extra" cancel forms need to have `method="post"` attributes

jseppi replied:
FYI an easy way to test that the fallback cancellation works is to temporarily comment out the `require('./modal-dialogs');` line in `/frontend/source/js/data-capture/index.js`

jseppi commented:
It would be super cool if someone (ahem, @toolness 😉 ) could whip up an auto crawler + screenshotter tool to make it easier to find design regressions like these

jseppi commented:
Also I know we haven't focused too much on responsive design, but although the cards generally work well on narrower displays, the top nav links don't:

<img width="639" alt="screen shot 2016-12-09 at 8 50 20 am" src="https://cloud.githubusercontent.com/assets/697848/21052870/a9ba7ac0-bdec-11e6-951e-8d08f9b00573.png">

ericronne commented:
Should we just hide the upload feature on mobile?
Can also probably kill "the" from the first one

hbillings commented:
The user-based pages are tricky, because there's no nav item to highlight. I threw the white card layout on them, but I think @ericronne could come up with a better solution.

The date ranges should really be side-by-side in both the narrow view and the wide view, but I think making it work on the narrow view requires changing the way the date widget labels the inputs (see #1079). 

I'll file a separate PR for the nav, as there's some other responsive stuff I need to tackle, also.

jseppi commented:
Is this ready to go again, @hbillings?

hbillings commented:
@jseppi yessir it is

jseppi commented:
The gap between the header and the start of the Data Explorer stuff seems very large to me (and is larger than in the screenshot in the root comment to this PR):

<img width="961" alt="screen shot 2016-12-13 at 10 46 59 am" src="https://cloud.githubusercontent.com/assets/697848/21149576/85e4aa02-c121-11e6-993f-0517ff09ae83.png">

Admin nav is not aligned to bottom of header, but I see you have a screenshot that shows it is properly aligned, so I wonder what is going on. I just pulled and completely rebuilt from this branch. Screenshot:

<img width="944" alt="screen shot 2016-12-13 at 10 49 47 am" src="https://cloud.githubusercontent.com/assets/697848/21149668/eb647dd0-c121-11e6-846b-c637d9ac7c92.png">

Step 3 Errors has floating `Cancel` link:

<img width="970" alt="screen shot 2016-12-13 at 10 52 17 am" src="https://cloud.githubusercontent.com/assets/697848/21149778/4b8ba990-c122-11e6-9f03-e865e38417b4.png">

Step 3 Errors has lots of spacing between page title and error message:

<img width="402" alt="screen shot 2016-12-13 at 10 53 09 am" src="https://cloud.githubusercontent.com/assets/697848/21149817/667b6b50-c122-11e6-98d1-d6df81f8182c.png">

Step 4 has floating `Cancel` link also:

<img width="975" alt="screen shot 2016-12-13 at 10 54 05 am" src="https://cloud.githubusercontent.com/assets/697848/21149856/856ff134-c122-11e6-9b05-e1fb1db01684.png">

Very minor, but the highlighted step in the step widget seems to be _slightly_ misaligned from the other steps. It is most evident in the final step:

<img width="439" alt="screen shot 2016-12-13 at 10 57 18 am" src="https://cloud.githubusercontent.com/assets/697848/21149972/0eea63a4-c123-11e6-91bd-091d71c30719.png">


Also without the modal javascript, the `Cancel` forms still do not work. They need the `csrf` tag, whoops :( Should be exactly as mentioned in https://github.com/18F/calc/pull/1078#discussion_r91593519:

```html
<form method="post">
  {% csrf_token %}
  <button type="submit" class="button-link" name="cancel" data-a11y-dialog-show="cancel-dialog" formnovalidate>Cancel this upload</button>
</form>
```

jseppi commented:
Also should probably have @ericronne do a final visual review for this PR also, yeah?

hbillings commented:
Okay, fixed most of these -- step 3 errors looks like this now:
<img width="1009" alt="screen shot 2016-12-14 at 9 45 01 am" src="https://cloud.githubusercontent.com/assets/509309/21191452/665f80d2-c1e2-11e6-905b-16ca2f434fd9.png">

The steps widget is going to be another PR because we're going to show/hide the labels like this:
<img width="239" alt="screen shot 2016-12-14 at 9 49 23 am" src="https://cloud.githubusercontent.com/assets/509309/21191523/9fd3be78-c1e2-11e6-9cb2-da8ff2de7be3.png">

jseppi commented:
Woohoo! All the stuff I noted earlier looks good!

Found one very minor issue on Step 2 of the Region 10 Bulk Upload (`/data-capture/bulk/region-10/step/2`), the buttons are kinda wonky:
<img width="669" alt="screen shot 2016-12-14 at 12 18 49 pm" src="https://cloud.githubusercontent.com/assets/697848/21195022/9eb7148a-c1f7-11e6-846f-5c7c583324df.png">

But feel free to address that in a separate PR.

jseppi commented on data_capture/templates/data_capture/step.html:
```diff
@@ -3,38 +3,58 @@
+      <div class="card">
+        <div class="card__content">
+          {% block steps %}
+          <div class="step-bar" aria-hidden="true">
```
might need to update the Style Guide with these step widget changes as well.

hbillings commented:
New header arrangement:
<img width="996" alt="screen shot 2016-12-14 at 3 34 50 pm" src="https://cloud.githubusercontent.com/assets/509309/21204113/f281b780-c212-11e6-8346-16d58a572eba.png">

hbillings commented:
Footer stuff
Logged out:
<img width="879" alt="screen shot 2016-12-14 at 4 42 00 pm" src="https://cloud.githubusercontent.com/assets/509309/21207058/112b7fb2-c223-11e6-94f5-5956b0ae9e51.png">

Logged in:
<img width="879" alt="screen shot 2016-12-14 at 5 26 41 pm" src="https://cloud.githubusercontent.com/assets/509309/21207082/31531c5a-c223-11e6-957e-da2c9b2378f3.png">

Home page:
<img width="876" alt="screen shot 2016-12-14 at 5 32 27 pm" src="https://cloud.githubusercontent.com/assets/509309/21207095/4f3acd80-c223-11e6-8feb-75cc887a7e42.png">

hbillings commented:
@jseppi Changed the bulk upload interface so it looks like this now:
<img width="631" alt="screen shot 2016-12-15 at 10 07 53 am" src="https://cloud.githubusercontent.com/assets/509309/21233985/6ded262a-c2ae-11e6-8760-7b57aff157ed.png">

toolness commented on data_capture/templates/price_lists/details.html:
```diff
@@ -28,79 +32,82 @@
+
+      {% if price_list_rows %}
+=======
+>>>>>>> develop
```
Oof, should probably remove these merge conflict annotations 😁 

> ![zoinks](https://cloud.githubusercontent.com/assets/124687/21262875/a61fb194-c361-11e6-9a7f-9ae19c8528e5.png)

toolness commented:
I think this is actually an old bug, but I figured I'd mention it here just in case... basically, the file upload widget's drag-over state where "Drop file here!" is displayed doesn't center the text vertically, which looks a bit weird to me:

> ![drop-file-here](https://cloud.githubusercontent.com/assets/124687/21263199/6e6577f0-c363-11e6-9ea3-5b66a7b09610.png)

Not that big a deal though, we can always file as a separate issue and fix later.

toolness commented:
I noticed that `/step/3/data` looks weird, but that's fine since we got rid of the view in #1110!

Hmm, it looks like uploading a price list with no valid rows has weird buttons at the bottom:

> ![want-to-try-that-again](https://cloud.githubusercontent.com/assets/124687/21263400/8c97c9fc-c364-11e6-8217-00425050d4b0.png)

Aside from the layout, that disabled "continue without these rows" button has a hover state--the text becomes black on hover--which it probably shouldn't have, right?

toolness commented:
Hmm, I also noticed that in the style guide, the disabled button example has black text and no hover state.  I suppose the white text makes it look more "disabled" but the black text does have better contrast--yet at the same time, since the button's unclickable, I'm not sure how important readability actually is.

Speaking of the style guide, it's still a bit of a mess, as documented in #1115... It would be great if that could be cleaned up a bit in this PR but I guess it is OK if not.

Actually, I could try taking a stab at cleaning up the style guide and submitting a PR to this PR if you want.

toolness commented on data_explorer/templates/index.html:
```diff
@@ -16,38 +16,31 @@
-      </div>
+  </div>
+  <div class="price-callout four columns">
+    <h3>With...</h3>
```
No `&hellip;`?! 😝

toolness commented on data_explorer/templates/about.html:
```diff
@@ -5,72 +5,89 @@
+  </div>
+  <div class="card">
+    <div class="card__content">
+      <h2>Where do the prices in this tool come from?</h2>
```
U know, Heydon Pickering might argue that the `.card` and `.card__content` styles here shouldn't have to be explicitly laid out in nested `<div>`s, and that the styling should just appear as part of `<h2>`s 😛  But that would be annoying to do in CSS and besides, it's not like HTML noobs are going to be editing this page. (Unless we want to migrate this content to be a [Django flatpage](https://docs.djangoproject.com/en/1.10/ref/contrib/flatpages/) at some point in the future, but that's the future!)
